File: 16 09 25/spiralmatrix.go
func spiralOrder(matrix [][]int) []int {
    if len(matrix) == 0 {
        return []int{}
    }

    res := []int{}
    top, bottom := 0, len(matrix)-1
    left, right := 0, len(matrix[0])-1

    for top <= bottom && left <= right {
        // left → right
        for j := left; j <= right; j++ {
            res = append(res, matrix[top][j])
        }
        top++

        // top → bottom
        for i := top; i <= bottom; i++ {
            res = append(res, matrix[i][right])
        }
        right--

        // right → left
        if top <= bottom {
            for j := right; j >= left; j-- {
                res = append(res, matrix[bottom][j])
            }
            bottom--
        }

        // bottom → top
        if left <= right {
            for i := bottom; i >= top; i-- {
                res = append(res, matrix[i][left])
            }
            left++
        }
    }

    return res
}

File: 16 09 25/spiralmatrix.rb
class Solution
  def spiral_order(matrix)
    res = []
    top, bottom = 0, matrix.length - 1
    left, right = 0, matrix[0].length - 1

    while top <= bottom && left <= right
      (left..right).each { |j| res << matrix[top][j] }
      top += 1

      (top..bottom).each { |i| res << matrix[i][right] }
      right -= 1

      if top <= bottom
        right.downto(left) { |j| res << matrix[bottom][j] }
        bottom -= 1
      end

      if left <= right
        bottom.downto(top) { |i| res << matrix[i][left] }
        left += 1
      end
    end

    res
  end
end

